From fc60c1cb6f8da397ed9fd393078c1e115cfb9f91 Mon Sep 17 00:00:00 2001
From: Harish Rachakonda <quic_rachakon@quicinc.com>
Date: Fri, 24 Nov 2023 18:27:16 +0530
Subject: [PATCH 1/1] wifi: mac80211: implement proper AP MLD HW restart

Previously, MLD restart is not implemented, so it only worked
for non-MLO. Add the needed code to restart an AP MLD correctly
after mode0 recovery.

Signed-off-by: Sathishkumar Muruganandam <quic_murugana@quicinc.com>
Signed-off-by: Harish Rachakonda <quic_rachakon@quicinc.com>
---
 net/mac80211/util.c | 46 +++++++++++++++++++++++++++++++++++++++++++--
 1 file changed, 44 insertions(+), 2 deletions(-)

--- a/net/mac80211/util.c
+++ b/net/mac80211/util.c
@@ -2481,6 +2481,35 @@ static int ieee80211_reconfig_nan(struct
 	return 0;
 }
 
+static void ieee80211_reconfig_ap_links(struct ieee80211_local *local,
+					struct ieee80211_sub_if_data *sdata,
+					u32 changed)
+{
+	int link_id;
+
+	for (link_id = 0; link_id < ARRAY_SIZE(sdata->link); link_id++) {
+		struct ieee80211_link_data *link;
+
+		if (!(sdata->vif.active_links & BIT(link_id)))
+			continue;
+
+		link = sdata_dereference(sdata->link[link_id], sdata);
+		if (!link)
+			continue;
+
+		if (rcu_access_pointer(link->u.ap.beacon))
+			drv_start_ap(local, sdata, link->conf);
+
+		if (!link->conf->enable_beacon)
+			continue;
+
+		changed |= BSS_CHANGED_BEACON |
+			   BSS_CHANGED_BEACON_ENABLED;
+
+		ieee80211_link_info_change_notify(sdata, link, changed);
+	}
+}
+
 int ieee80211_reconfig(struct ieee80211_local *local)
 {
 	struct ieee80211_hw *hw = &local->hw;
@@ -2754,6 +2783,12 @@ int ieee80211_reconfig(struct ieee80211_
 		case NL80211_IFTYPE_AP:
 			changed |= BSS_CHANGED_P2P_PS;
 
+			if (sdata->vif.valid_links)
+				ieee80211_vif_cfg_change_notify(sdata,
+								BSS_CHANGED_SSID);
+			else
+				changed |= BSS_CHANGED_SSID;
+
 			if (sdata->vif.bss_conf.ftm_responder == 1 &&
 			    wiphy_ext_feature_isset(sdata->local->hw.wiphy,
 						    NL80211_EXT_FEATURE_ENABLE_FTM_RESPONDER))
@@ -2762,8 +2797,13 @@ int ieee80211_reconfig(struct ieee80211_
 			if (sdata->vif.type == NL80211_IFTYPE_AP) {
 				changed |= BSS_CHANGED_AP_PROBE_RESP;
 
-				if (rcu_access_pointer(link->u.ap.beacon))
-					drv_start_ap(local, sdata, link->conf);
+				if (sdata->vif.valid_links) {
+					ieee80211_reconfig_ap_links(local, sdata, changed);
+					break;
+				}
+
+				if (rcu_access_pointer(sdata->deflink.u.ap.beacon))
+					drv_start_ap(local, sdata, sdata->deflink.conf);
 
 				ieee80211_vif_cfg_change_notify(sdata, BSS_CHANGED_SSID);
 			}
@@ -2772,8 +2812,7 @@ int ieee80211_reconfig(struct ieee80211_
 			if (sdata->vif.bss_conf.enable_beacon) {
 				changed |= BSS_CHANGED_BEACON |
 				BSS_CHANGED_BEACON_ENABLED;
-				ieee80211_link_info_change_notify(sdata, link, changed &
-								  ~BSS_CHANGED_IDLE);
+				ieee80211_bss_info_change_notify(sdata, changed);
 			}
 			break;
 		case NL80211_IFTYPE_NAN:
